### DISCUSSION TITLE
Advice on style training

### DISCUSSION THREAD

**Discussion by Clement-Lelievre** [huggingface/diffusers#10787] [Q&A]
Hi all,

I'm looking for advice on how to best train a LoRA on the following task: the LoRA should learn to represent black and white wireframe drawings seen from top (orthogonal view ie objects seen from top, no perspective).
I attach one example of image from my dataset depicting a desk in the above mentioned style.

So far I've mostly tried fine-tuning flux dev, varying lr, rank etc. to see what I can get; and the LoRA partially learns but definitely isn't great.

Do you have any recommendations of model/hyperparams? (if this task is feasible at all)

Thanks vm!
Happy to provide more details if needed
![dataset_desk_14](https://github.com/user-attachments/assets/d7b600d6-6c50-4169-9164-d7d7025e4fdd)

(maybe @asomoza ?)

**Comment by asomoza** [comment]
can you post an image generated with your lora?

It's probably doable with any model but it depends on your definition of 'isn't great'. For example, there's a lora for ikea instructions trained in the base SDXL model which is fine if you don't consider the text.

About your decision to use Flux, I haven't tried to train a lora on it yet but for what I've seen and tested, Flux looses a lot of the quality when you finetune it, the loras I tested work fine but they degrade a lot the other stuff and also the model isn't that great for lines (for example anime) and it's really overtrained in professional photos with bokeh.

For example, take this [lora for SDXL](https://civitai.com/models/434354/wire-frame) and this [one for Flux](https://civitai.com/models/861126/wireframe-3d-asset-mesh) which are for 3D wireframes, if you ask me, based in the demo images, I think the SDXL one is better.

Also how big if your dataset? Depending on the size, you will probably need to finetune a model that's really over trained in lines like the anime models. This probably needs a bigger dataset than a normal character lora.

I really can't help with the hyperparameters for Flux, but one good start would be to train the whole model and then extract loras in different ranks to test which one works better for your use case. Also as a good default start, you should use the same LR than what the base model used.

**Reply by Clement-Lelievre** [reply]
hi @asomoza 

Thanks for the quick and detailed reply.

> Flux looses a lot of the quality when you finetune it, the loras I tested work fine but they degrade a lot the other stuff and also the model isn't that great for lines

this is consistent with my best checkpoint's output (see below)

> Also how big if your dataset? Depending on the size, you will probably need to finetune a model that's really over trained in lines like the anime models. This probably needs a bigger dataset than a normal character lora.

I started with the advised dataset (10 to 20) size so went for 20 image/caption pairs. I tried detailed captions but it was unclear the lora actually learnt, so then I opted for minimal, single-word caption this way I could immediately see if the LoRA learnt , by comparing inferences with respective prompts "sink" and "TOK sink" for example (TOK being obviously the activation token in my case)
I can get a bigger dataset maybe up to around 100.

> one good start would be to train the whole model and then extract loras in different ranks to test which one works better for your use case.

This is exactly what I did, I started with rank 64 and 6000 steps , leaving the recommended lr, saving a checkpoint every 100 steps and inferring with them, and that's how I noticed that the 700-step checkpoint is by far the best, as it's basically the only one that picks up orthogonal view, or close to. However the dark lines are a bit lost. Then I noticed rank 16 was good enough. See examples below.

**results:**
"plant"
<img src="https://github.com/user-attachments/assets/1637ec89-80d5-4c32-aa1f-4209dffc0a53" alt="image" width="200" height="200">

"car"
<img src="https://github.com/user-attachments/assets/3abb8bd8-5de7-4158-b397-d3a7131faf4b" alt="image" width="200" height="200">


"sofa"
<img src="https://github.com/user-attachments/assets/236b0ffc-a1a6-43cc-b1ea-9deef04a3706" alt="image" width="200" height="200">

**Reply by asomoza** [reply]
you seem to be in the right track, probably you just need a bigger dataset, also maybe you can try increasing the contrast of the images to make the darker lines more prominent, sometimes the model doesn't see the things the same as we do, so for example, what looks to us like a black line, the model may interpret it as a light gray line.

As a suggestion and because I'm kind of curious, maybe you can try doing a test run with an illustrious finetuned SDXL model, those model are really impressive if you don't take into account that they're mostly trained for nsfw, also there's a new animagine model that has really detailed line generations.

**Reply by Clement-Lelievre** [reply]
thanks
I'm not a big fan of using a model "mostly trained for nsfw" and I don't understand your point :) ?
if you can suggest a model I may give it a try

**Reply by asomoza** [reply]
> I'm not a big fan of using a model "mostly trained for nsfw" and I don't understand your point :) ?

Explaining it In a easy to understand way, those models are trained 100% on anime and drawings (lots of lines), which is exactly what you're doing, the only difference is that the training is done with nsfw content, so you will only need it to learn your specific type of drawings and not how to draw.

Probably you can achieve the same with any model, but it will take longer to train and a much bigger dataset specially for one that it's mostly trained on real photos like Flux.

> if you can suggest a model I may give it a try

sadly I can't, I just test them when there's a hype on them to have the knowledge on the SOTA for anime and drawings and to have a base line for newer models. I don't use them beside for initial testing mostly because I like a lot more doing real images. But basically any model trained with lines will do a better job than one trained with photos.

**Reply by Clement-Lelievre** [reply]
Maybe I'll try, but remember that I am trying to teach several aspects of my dataset images, including orthographic view; and I assumed that flux, with its bigger architecture can understand this better than SDXL-derived models. that's why I started with flux. I suspect that both on SDXL and FLUX training dataset, most images are not orthographic and have perspective.

I noticed that objects as mundane as a chair are difficult to render from top (I mean, really, like 90 degree angle, right above the object) at least to flux-dev.

Curiously the 700 steps LoRA (the one I used to generate the examples above)  hits a sweet spot as it manages orthographic views on a large variety of prompts including prompts unseen at training time, but all other checkpoints fail to perform orthographic representations! (at least with my hyperparams)

**Reply by asomoza** [reply]
that's how training is, it's not an exact science and it involves a lot of trial and error. 

I like to train more on epochs than steps, but I've got a lot of cases where an earlier epoch was better than most of the later ones until there's another one in the later epochs that also works, to me, it seems your lora it's still under trained even if you think the 700 step one works better. There also some other strategies you can apply, like lowering the LR after the 700 steps mark.

Also it seems you are underestimating training and overestimating the model capabilities, as I told you before, Flux degrades quite a lot with finetuning, so you'll need to mitigate this with something, in your case, since you want to preserve orthographic views or for it to preserve the angles and camera views, you should also have those in your dataset, specially if you're using dreambooth.

For example, since you're using a token I guess you're using dreambooth, and if most of your dataset has top views, the model will learn that your token means top view, it doesn't matter if it understands other views, just having a couple of images with orthographic views won't mitigate this.

**Reply by Clement-Lelievre** [reply]
yes I'm using Dreambooth and LoRA

> it seems your lora it's still under trained even if you think the 700 step one works better

at least on the intermediate inferences made with the checkpoints every 100 steps (up to 6000 steps!), like I said the 700 steps is by far the best-performing but yeah there may very well be a superior set of params

> There also some other strategies you can apply, like lowering the LR after the 700 steps mark

yes will try this thks

> since you want to preserve orthographic views or for it to preserve the angles and camera views, you should also have those in your dataset, specially if you're using dreambooth.

All of my dataset images are seen from top with orthographic view, in exactly the same style as the initial image I shared

**Reply by asomoza** [reply]
I misunderstood what you said, I read it again and you were referring to the base training, don't really know about Flux but SDXL can do top down views without a problem if you fine tune it for it. 

I did a quick search for a lora and I could generate these images really easy:

|||||
|---|---|---|---|
|![20250214104848](https://github.com/user-attachments/assets/f0fe5cb6-6acd-4779-ad9c-56351d6b3ab2)|![20250214105030](https://github.com/user-attachments/assets/7ac4917d-12d9-4a59-b630-98e76bc8ce17)|![20250214105107](https://github.com/user-attachments/assets/20f36b01-39f3-453a-82e8-1df6f38c29f2)|![20250214105219](https://github.com/user-attachments/assets/f9840948-ab38-4c78-90f5-83afa23d2342)|

Didn't see anything similar in Flux, just some people that trained some top down map generators.

What you're trying to do shouldn't be that hard, you're just training with grayscale images and with straight lines, a lot easier than some of the other finetuning's like those weird styles like yarn or spaghetti, also I saw someone that trained a lora with dotted lines for drawing which seems a lot more complex to train than yours.

My guess is that since you're fixated in using Flux, you will need to teach it everything, first the top view, then the line drawings and then the conversion of the objects between for example, a front view and a top view. 

Probably you won't be able to do this with dreambooth and will need to train a full lora or model for this, also take into account, that people mostly finetune loras with Flux for portrait photos of a person with a very simple pose, that is probably for a reason.

**Reply by Clement-Lelievre** [reply]
thanks for the reply. haha don't worry I'm not fixated with using flux, if another model is better suited to the task then I'll go for it.
So you're saying you found an existing SDXL LoRA that generated the top view images you shared?
I will look into dreambooth sdxl lora training

**Reply by asomoza** [reply]
yeah, [this one](https://civitai.com/models/815019/topdown-map-assets-sci-fi) but as I told you, it's for a model that's not that safe to use but that is only trained with lines.

I used to never look at those models but I read and then tested (and corroborted) that they're really good for a lot of other stuff because of the same training. The problem is that even if you don't prompt it, you can sometimes get some nsfw results, so if that bothers you, you will need to train or find a model trained with lines that is sfw but because of the same, they lack the understanding of these newer models.

Same problem as base models with anatomy, SDXL, SD3, an all the newer models are really bad with anatomy because they weren't trained with "good anatomy" to avoid the backslash and to be stigmatized as nsfw.

**Reply by Clement-Lelievre** [reply]
update: I fine-tuned SDXL (still with Dreambooth LoRA, using [this diffusers script](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_sdxl.py)) with the default params, and while the LoRA quickly learnt the style (white background, thin dark lines, wireframe style) it struggles with learning orthogonality, specifically drawing an object as seen from top (like the first image I posted, which comes fro my training dataset). Flux struggled with this too.
I tried to play with the hyperparams a bit but got no improvement.
Not sure if it is realistic to teach this "concept" (I guess it's a concept?) of orthogonality in a Dreambooth LoRA fine-tuning, I don't have enough experience of it.

**Reply by asomoza** [reply]
yeah, this probably is not something you can teach the model with just dreambooth and more like a full lora or finetune.

I also have an update, I've seen a lot more people producing Flux loras lately (probably because of [Flux Gym](https://github.com/cocktailpeanut/fluxgym) and there's been a couple of new ones that are using styles which seems good, at least in the demo images, so this makes me think that you will be able to train your idea with Flux.

Probably same as you have guessed already, not with dreambooth and with a larger dataset.

**Comment by Clement-Lelievre** [comment]
hey @asomoza 

I wanted to update you in case you're interested: I've come up with a way to remove one constraint in my training (the orthographic representation). So now, I am trying to teach the model the style shown on the image I displayed in my first message above. I have a balanced mix of all views in my dataset (diverse objects, shown in front, side or top view).

(I will then load these lora weights and use a controlnet in order to get this style AND following the structure of a depth map, but that's another story)

I couldn't make flux learn it, now [trying with SDXL](https://wandb.ai/rayon/dreambooth-lora-sd-xl/reports/validation-25-03-10-17-11-09---VmlldzoxMTcyODU4OA?accessToken=7p8b9hbr6nqsdhie3tbvdymflwlw9j12dux6i1ddznd0jsk3p78mdkfrobn15lkf). Using [this script](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_sdxl.py) with hyperparams defaults for now. As you can see, the training is happening to some degree, as validation inferences show a drawing style, and dark thick edges, but their style is far from the wireframe, minimalist drawing style of the dataset

**Reply by asomoza** [reply]
@Clement-Lelievre nice!, I still don't get the complete scope of what you're trying to do though, if you're going to use a controlnet for the orthographic representation, for me, that's all that there's to it so what do you still need to train the model?

If I grab your image from the first post and just use a controlnet with it:

|lineart|special blend (edge)|
|---|---|
|![20250310184826](https://github.com/user-attachments/assets/57ff068d-cf1f-4246-8df1-76b78e9dfe3d)|![20250310184545](https://github.com/user-attachments/assets/823d5bf3-f6bf-4e9e-bb64-a8474cb260ef)|

Those results are even without playing with the brightness or contrast which probably will give better results.

For me the style it's just line drawings, what is that you need the model to train? For me the `orthographic representation` is pretty much the hard part of what you're trying to do and it's a cool project since it's something that humans take a long time to do.

Probably I'm still missing something on what you're doing but anyways, glad you got some progress and I think it's a really good learning experience and a good project too.

**Reply by Clement-Lelievre** [reply]
hi @asomoza 
to clarify

in my use case, I will get depth maps (or normal maps if I want to), and from them, I need to produce a raster image that has:
- the structure from the abovementioned  depth/normal map, which will act as control image
- the style from the dataset image I shared

I do not have access to a dataset of depth map/target image pairs so I cannot train a Controlnet.

Therefore, what I did to try and accomplish this task is:
- finetune SDXL txt2img with Lora Dreambooth (using a dataset of 30 image-only mixed views, no captions) ([I went up to 4000 steps](https://wandb.ai/rayon/dreambooth-lora-sd-xl/reports/test-25-03-11-10-32-33---VmlldzoxMTc0MDQwMg?accessToken=5sctdwu7zicbg7dmjp0vvy47m703pb49cvn2eqq4pgbz3eff2y58fg7ag2y29slj))
- load `StableDiffusionXLControlNetPipeline` and load the LoRA weights from the best checkpoint, to infer


While it works in the sense that:
- there is definitely some training happening, the style is learnt (thicker dark contours, light gray inner edges, white background etc.) but not perfectly
- control image stricture is well respected
- the LoRA loading does not break the controlnet weights

it is still highly sensitive to params, and far behind expectations visually, see images

![image](https://github.com/user-attachments/assets/b100466b-ce1e-4f6a-9729-b307ffc4f7a8)
![image](https://github.com/user-attachments/assets/08cec4d5-16b8-4275-9735-4e4a456ee56f)


The key thing is, I need consistency between views of the same object (eg a car seen from front or from the side), and right now, even if the structure is well respected, the style is not learnt well enough and therefore there is no visual consistency across views. Likewise, just changing the seed produces a noticeable variation of the same style while I need the model to really enforce a single style

**Comment by Clement-Lelievre** [comment]
hey @asomoza 

a few months after, this issue is still current :) 
I was using `nanobanana` to generate images of objects from top view in a technical drawing style, and even this praised model struggles to produce top views. It often produces perspectives instead, which are unusable for me because my use case is: 
`text prompt > image > vector (SVG) > insert vector on architect's floorplan
`

I see several possible reasons for the model's inability to generate accurate top views, including:

- Lack of top views in the training set (even for us humans, some objects are not trivial to draw from an orthographic top view)
- The prompt is almost contradictory, because many objects just look like a rectangle/oval from above (e.g., table, stool, fridge, etc.), and it seems like the model absolutely wants to show the object's features 

See example images below of a washing machine: front and side views are fine, but then the top view is incorrect, it's like the model desperately wants to show it's a washing machine, while a mere rectangle would be fine instead of this

<img width="256" height="256" alt="front" src="https://github.com/user-attachments/assets/2724fa55-5c6f-4a90-aa49-ca1c9a87b54b" />
<img width="256" height="256" alt="side" src="https://github.com/user-attachments/assets/98dc2f80-a899-4f2a-a8f1-beece50720ba" />
<img width="256" height="256" alt="top" src="https://github.com/user-attachments/assets/090e82e1-2e4a-4a57-a54c-15606c2a1427" />


When the object prompted is well-known from top view (meaning well-known to us and to the training sets), like say a pool table, the model does fine. But when it is an object unfrequently depicted from top view (eg a fridge) then the model struggles. Sometimes it even generates a front view instead of a top view, as if it was completely giving up on the idea of respecting the prompt.

I want to ask you, in light of the past months of research, do you see any way forward to tackle this issue? I was thinking of fine-tuning flux kontext, wdyt?

thank you for your answer!

**Reply by asomoza** [reply]
I think this is a general problem with the training and models, if nano banana couldn't do it, probably no other model could without a lot of training. There's the new qwen-image-edit model which accepts multiple images, that model has a really good prompt adherence so maybe you can try it with a really good prompt and a couple of images, like for example, the source image and a another one showing how to do it.

Since your original post, I haven't seen any other lora or generation that does something similar, top down views are really hard to train and to understand to diffusion models it seems.
